xwayland/xwm.c

```diff
@@ -1246,6 +1248,11 @@ static void xwm_handle_focus_in(struct wlr_xwm *xwm,
 	struct wlr_xwayland_surface *requested_focus = lookup_surface(xwm, ev->event);
 	if (xwm->focus_surface && requested_focus &&
 			requested_focus->pid == xwm->focus_surface->pid) {
+		// We let the application override focus immediately after our focus change
+		// request, but not later
+		if (ev->sequence != xwm->last_focus_seq) {
```

Comment:
Are there any cases where this is `==`? It feels to me like this is basically an `if (true)`:

- `xwm_send_focus_window()` calls `xcb_set_input_focus()` and records the sequence number
- Immediately afterwards it calls `xcb_configure_window()`

Most likely the two requests are sent to the X11 server in a single call to `write()`, thus it it basically impossible for anything to cause a focus change in between the two requests. (and the sequence number of an event is the sequence of the last request that was handled by the server, so should always be the newer one.)

Reply:
Yes, there are actually quite a lot of cases where this is `ev->sequence == xwm->last_focus_seq`. Just add a `wlr_log` after this `if`, and run steam.

Unfortunately, I can't disprove your logic either :/

Reply:
Okay, thanks for checking. In other news, I have no idea what is happening here. :-)

Reply:
Well, I do have an idea what is going on. Of course, `xcb_set_input_focus()` causes a focus change! That's its whole job. That focus change causes a `FocusIn` event to be sent, which will satisfy the `sequence` condition.

I guess that means that this patch basically "kills" what this code is supposed to do: Allowing apps to change the focus themselves. From the comment above, that should bring back https://github.com/swaywm/sway/issues/1865.

Reply:
> From the comment above, that should bring back [swaywm/sway#1865](https://github.com/swaywm/sway/issues/1865)

Except that it doesn't. I have verified steam and it works as expected, even with this patch.

Also what you said cannot be true - otherwise, the current code in master would trigger an infinite loop with Xwayland, because we'd be continuously sending focus in response to focus in event.

Reply:
A `FocusIn` event is only generated when the focus actually changes. So, setting the focus to whatever already has the focus is a no-op. That would stop the loop.

Thanks for testing that Steam still works and sorry that I am always coming up with broken theories.